src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 	uint32 rangeTableId = 1;
 	Oid insertRelationId = insertRte->relid;
 	Var *insertPartitionColumn = PartitionColumn(insertRelationId, rangeTableId);
-	bool partitionColumnsMatch = false;
 	Query *subquery = subqueryRte->subquery;
+	bool targetTableHasPartitionColumn = false;
 
 	foreach(targetEntryCell, query->targetList)
 	{
 		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		List *insertTargetEntryColumnList = pull_var_clause_default((Node *) targetEntry);
+		int insertTargetEntryColumnCount = list_length(insertTargetEntryColumnList);
+		Var *insertVar = NULL;
+		AttrNumber originalAttrNo = InvalidAttrNumber;
+		TargetEntry *subqeryTargetEntry = NULL;
 
-		if (IsA(targetEntry->expr, Var))
+		/*
+		 * We only consider target entries that include a single column. Note that this
```

Comment:
In what cases could the target entry contain more than one column? I don't quite understand this comment… This target list is the list of the `INSERT INTO`, not the one from the `SELECT` subquery, right?

I could see how the target entry might be a field or array reference, but I don't quite get what this check is for… maybe I'm misunderstanding something. Do you have an example of a query that triggers this? I checked the coverage report and it appears this case isn't tested, which makes me wonder what it's for.